### PR TITLE
[SPARK-50513][SS][SQL] Split EncoderImplicits from SQLImplicits and provide helper object within StatefulProcessor to access underlying SQL Encoder related implicit functions

### PR DESCRIPTION
**Pull request by anishshri-db**
### What changes were proposed in this pull request?
Split EncoderImplicits from SQLImplicits and provide helper object within StatefulProcessor to access underlying SQL Encoder related implicit functions

### Why are the changes needed?
Without this, we cannot handle the implicit encoder APIs on the executor. We would run into a NPE since the spark session is not available on the executors. One option is to pass the `SparkSession` or `SQLImplicits` directly to the `StatefulProcessor`. However, this risks exposing some methods relying on `SparkSession` again.

```
Caused by: Task 11 in stage 4.0 failed 4 times, most recent failure: Lost task 11.3 in stage 4.0 (TID 57) (10.68.181.194 executor 2): java.lang.NullPointerException: Cannot invoke "org.apache.spark.sql.SparkSession.implicits()" because the return value of "$line94e51c7e8a51415d827a613fcad00cc19.$read$$iw$$iw.spark()" is null
	at $line94e51c7e8a51415d827a613fcad00cc152.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$FruitCountStatefulProcessor.init(command-412580759966543:18)
	at org.apache.spark.sql.execution.streaming.TransformWithStateExec.processDataWithInitialState(TransformWithStateExec.scala:732)
	at org.apache.spark.sql.execution.streaming.TransformWithStateExec.$anonfun$doExecute$2(TransformWithStateExec.scala:625)
```

Hence, we do 2 things:
- split the encoder related functions that don't rely on `SparkSession` into a separate trait
- expose a helper `implicits` object within the `StatefulProcessor` interface that users can import directly, providing access to only the necessary functions

### Does this PR introduce _any_ user-facing change?
Yes

### How was this patch tested?
Added unit tests

```
[info] Run completed in 9 minutes, 30 seconds.
[info] Total number of tests run: 128
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 128, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
```

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by anishshri-db** [comment]
cc - @cloud-fan @HeartSaVioR - PTAL, thx !

File: sql/api/src/main/scala/org/apache/spark/sql/streaming/StatefulProcessor.scala [hunk 1]
```diff
@@ -20,16 +20,27 @@ package org.apache.spark.sql.streaming
...
+ * Users can also explicitly use `import implicits._` to access the EncoderImplicits and use
+ * the state variable APIs relying on implicit encoders.
  */
 @Experimental
 @Evolving
 private[sql] abstract class StatefulProcessor[K, I, O] extends Serializable {
 
+  // scalastyle:off
+  // Disable style checker so "implicits" object can start with lowercase i
+  object implicits extends EncoderImplicits {
```
Comment (HeartSaVioR) [inline sql/api/src/main/scala/org/apache/spark/sql/streaming/StatefulProcessor.scala]: nit: we don't need an empty body; let's remove {}

**Review by HeartSaVioR — COMMENTED** [review]
The code change looks good to me. Have you tried with having test(s) leveraging `local-cluster`? IIUC, the issue was only happening with distributed setup, hence we'd want to test the case if feasible.

File: sql/api/src/main/scala/org/apache/spark/sql/streaming/StatefulProcessor.scala (hunk 1 above)
Comment (HeartSaVioR) [reply sql/api/src/main/scala/org/apache/spark/sql/streaming/StatefulProcessor.scala]: It's still valid, let's see {} is necessary.

Comment (anishshri-db) [reply sql/api/src/main/scala/org/apache/spark/sql/streaming/StatefulProcessor.scala]: Done - removed it

**Comment by anishshri-db** [comment]
> The code change looks good to me. Have you tried with having test(s) leveraging local-cluster? IIUC, the issue was only happening with distributed setup, hence we'd want to test the case if feasible.

In the test context, atleast I'm not able to compile if we don't pass the implicits in some form. But let me add a `local-cluster` suite anyway - to ensure that the serialization path is hit and working as expected

**Comment by HeartSaVioR** [comment]
> to ensure that the serialization path is hit and working as expected

Yes please. This is the path we had an issue I assume.

**Comment by anishshri-db** [comment]
> Yes please. This is the path we had an issue I assume.
Done - added the suite

**Review by cloud-fan — APPROVED** [review]

**Review by HeartSaVioR — APPROVED** [review]
+1

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Pull request closed without merging**